### ISSUE TITLE
strip_url_params.py : BUG in function strip_url_params3(...)

### ISSUE THREAD

**Issue by christianbender** [keon/algorithms#256]
The file ```strip_url_params.py``` is buggy. Because one times the tests run successful and a other times it fails. I comment out the test for this file.  

***important:*** The bug is in the function ```strip_url_params3``` of the file above.  

@SaadBenn Can you look in it? The project is from you.

**Comment by SaadBenn** [comment]
Yeah sure, currently I'm a bit busy. Is it okay if I take my time with it? I have some huge assignments atm that I need to complete and that's the reason I have not been so active lately. @christianbender

**Comment by danghai** [comment]
Could you elaborate the error you got?

**Comment by christianbender** [comment]
@danghai The error message of the failed test looks so
> line 330, in test_strip_url_params3
    self.assertEqual(strip_url_params3("www.saadbenn.com?a=1&b=2&a=2"), "www.saadbenn.com?a=1&b=2")
AssertionError: 'www.saadbenn.com?b=2&a=1' != 'www.saadbenn.com?a=1&b=2'
> - www.saadbenn.com?b=2&a=1
> ?                     ----
> + www.saadbenn.com?a=1&b=2
> ?                  ++++

**Comment by SaadBenn** [comment]
I'll step through it today once I get off work. @christianbender

**Comment by christianbender** [comment]
@SaadBenn Thanks

**Comment by programatt** [comment]
@christianbender @SaadBenn I took a shot at it. I think it just needs to use an OrderedDict instead of a defaultdict so that the looping of the keys is deterministic. See PR #367

**Comment by programatt** [comment]
@danghai Uh oh, looks like it didn't fix it. The build failed.

**Comment by danghai** [comment]
@programatt Yeah it is fail in python 3.4 and 3.5. Could you re-check them?

**Comment by programatt** [comment]
@danghai it's my fault. I was looking at strip_url_params1 not strip_url_params3. Looks like the implementation of `urllib.parse.urlencode` or urllib.parse.parse_qs` might have changed between python 3.4 and 3.6. I'll take a look.

**Comment by programatt** [comment]
Ah, it appears to be the result of the new implementation of dictionaries in 3.6+ preserving key order by default. [Stackoverflow Question about Dictionary Key Order in 3.6+](https://stackoverflow.com/q/39980323). So the test is using a query string of `?a=1&b=2&a=1` and expecting `?a=1&b=2`. `?a=1&b=2` and `?b=2&a=1` are in fact the same query string from the perspective of any library reading an http request. So either we can update the test to expect a different string when in python 3.6+ or we can sort the keys in the implementation. Which one seems "cleaner" to everyone? It says in the stackoverflow answer to not rely on the key order changing in future versions of python. So I would vote for doing the sort even if it means the implementation is a little *uglier*.

**Comment by goswami-rahul** [comment]
@danghai didn't the test pass with `OrderedDict`?

**Comment by keon** [comment]
Fixed in PR #2719 — tests uncommented and passing. The bug was a Python 3.5-era dict ordering issue; `parse_qs` returns consistently ordered results in Python 3.7+.
